src/snowflake/cli/plugins/nativeapp/manager.py

```diff
@@ -773,6 +777,52 @@ def get_events(
         except ProgrammingError as err:
             generic_sql_error_handler(err)
 
+    def stream_events(
+        self,
+        last: int,
+        interval_seconds: int,
+        record_types: list[str] | None = None,
+        scopes: list[str] | None = None,
+    ) -> Generator[dict, None, None]:
+        try:
+            events = self.get_events(
+                record_types=record_types, scopes=scopes, last=last
+            )
+            yield from events  # Yield the initial batch of events
+            last_event_time = events[-1]["TIMESTAMP"]
+
+            while True:  # Then infinite poll for new events
+                time.sleep(interval_seconds)
+                previous_events = events
+                events = self.get_events(
+                    since=last_event_time, record_types=record_types, scopes=scopes
+                )
+                if not events:
+                    continue
+
+                yield from _new_events_only(previous_events, events)
+                last_event_time = events[-1]["TIMESTAMP"]
+        except KeyboardInterrupt:
+            return
+
+
+def _new_events_only(previous_events: list[dict], new_events: list[dict]) -> list[dict]:
+    # Returns the events in new_events that were emitted after the ones in previous_events
+    # Handles duplicates by checking for an overlap until one can't be found anymore
+    overlap_amount = 1
+    last_overlap_found_at = 0
+    while overlap_amount <= min(len(new_events), len(previous_events)):
+        # Check if end of previous_events overlaps with start of new_events
+        # This is O(n) but from the SQL filters and ordering used to generate
+        # both lists of events, we know that the only overlap would be for events
+        # that happened in the exact same microsecond (usually only 1 or 2, rarely more)
+        if previous_events[-overlap_amount:] == new_events[:overlap_amount]:
+            last_overlap_found_at = overlap_amount
+        elif last_overlap_found_at:
+            break
```

Comment:
Is it guaranteed that events with the same timestamp arrive in the same order persistently? If not, it might be safer to check all events with the same microsecond and not stop at the first found.

Reply:
yeah that's a good point, since the query has only `order by timestamp asc` there's no guarantee that the rows with the same timestamp will be ordered the same on the next call. i'll update the logic to handle this